buffer/src/main/java/io/netty/buffer/UTF8Decoder.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.buffer;
+
+import java.io.DataInput;
+import java.io.IOException;
+
+import io.netty.util.ByteProcessor;
+import io.netty.util.internal.MathUtil;
+import io.netty.util.internal.StringUtil;
+
+/**
+ * A fast utf decoder implemented based on http://bjoern.hoehrmann.de/utf-8/decoder/dfa/
+ */
+public final class UTF8Decoder {
```

Comment:
I think it could be very useful for apps, however we could consider putting it in the 'common' module and moving the UTF8Decoder.decode(ByteBuf) method to ByteBufUtil.